src/include/storage/undo_record.h

```diff
@@ -165,30 +215,35 @@ class UndoRecord {
    * @param timestamp timestamp of the transaction that generated this UndoRecord
    * @param slot the TupleSlot this UndoRecord points to
    * @param table the DataTable this UndoRecord points to
+   * @param txn the transaction this UndoRecord belongs to
    * @param redo the redo changes to be applied
    * @return pointer to the initialized UndoRecord
    */
   static UndoRecord *InitializeUpdate(byte *const head, const transaction::timestamp_t timestamp, const TupleSlot slot,
-                                      DataTable *const table, const storage::ProjectedRow &redo) {
+                                      DataTable *const table, transaction::TransactionContext *const txn,
+                                      const storage::ProjectedRow &redo) {
     auto *result = reinterpret_cast<UndoRecord *>(head);
 
     result->type_ = DeltaRecordType::UPDATE;
     result->next_ = nullptr;
     result->timestamp_.store(timestamp);
     result->table_ = table;
     result->slot_ = slot;
+    result->txnptr_.Put(txn);
 
     ProjectedRow::CopyProjectedRowLayout(result->varlen_contents_, redo);
 
     return result;
   }
 
  private:
+  friend class GarbageCollector;
```

Comment:
Is it necessary to add GarbageCollector as friend class? Add a public function to provide txnptr_ may be better?

Reply:
We would need getters and setters for all 7 fields of undo record class. Moreover, adding get and set would allow everyone to modify the class.